src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -3052,6 +3054,41 @@ class RGWObjFetchCR : public RGWCoroutine {
   }
 };
 
+RGWCoroutine *RGWDefaultDataSyncModule::fail_object_replication_status(const DoutPrefixProvider *dpp,
+                                                                       RGWDataSyncCtx *sc, rgw_bucket_sync_pipe &sync_pipe,
+                                                                       rgw_obj_key &key)
```

Comment:
because the source object isn't necessarily immutable, there's a potential race between source object overwrites and this replication status update. we might want to include an `If-Unmodified-Since` header with this request similar to the one we pass with the GET request from `fetch_remote_obj()`

Reply:
Would it be necessary to call `RGWStatRemoteObjCR()` to retrieve the mtime of the source object, or might I be overlooking something?

for `fetch_remote_obj()` we check the local dest object to fill `If-Modified-Since`.

Reply:
I have implemented one for set_attrs.

Reply:
the calls to `set_replication_status_header()` from `RGWGetObj::verify_permission()` are missing similar protection: while sync's `GetObject` requests do contain `If-Unmodified-Since`, that's not enforced until later in `RGWGetObj::execute()`

you've added the PATCH method to handle lack of destination bucket permissions - maybe we should handle the source bucket permissions the same way? the replication status updates may be worth splitting out to a separate pr

Reply:
> you've added the PATCH method to handle lack of destination bucket permissions - maybe we should handle the source bucket permissions the same way?

agree, and good point. We are already calling `fail_object_replication_status()` on `-EACCES` which is also the result of the source object fetch. So we are actually doing it twice, and therefore, we can drop it and leave it to patch.

> the replication status updates may be worth splitting out to a separate pr

yeah, I'd love to. I was drafting here mostly as testing is heavily reliant on qa commits from this PR https://github.com/ceph/ceph/pull/61828. @cbodley, by any chance, do you think we can have a qa on that PR sooner so I can split it into two PRs more easily?